src/handshake/http.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT
+// license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. All files in the project carrying such notice may not be copied,
+// modified, or distributed except according to those terms.
+
+//! This module exposes a utility method and a couple of related types to
+//! make it easier to upgrade an [`http::Request`] into a Soketto Socket
+//! connection. Take a look at the `examples/hyper_server` example to see it
+//! in use.
+
+use super::SEC_WEBSOCKET_EXTENSIONS;
+use crate::handshake;
+use crate::handshake::server::ServerConfiguration;
+use http::{header, HeaderMap, Response};
+use std::convert::TryInto;
+
+/// An error attempting to upgrade the [`http::Request`]
+pub enum UpgradeError {
+	/// The [`http::Request`] provided wasn't a socket upgrade request.
+	NotAnUpgradeRequest,
+	/// A [`handshake::Error`] encountered attempting to upgrade the request.
+	HandshakeError(handshake::Error),
+}
+
+impl From<handshake::Error> for UpgradeError {
+	fn from(e: handshake::Error) -> Self {
+		UpgradeError::HandshakeError(e)
+	}
+}
+
+/// This is handed back on a successful call to [`upgrade_request`].
+pub struct Upgraded {
+	/// This should be handed back to the client to complete the upgrade.
+	pub response: http::Response<()>,
```

Comment:
We are actually only returning a set of headers, so I could conceivably return just those, and let the caller construct a `Response` themselves. Headers would be a more precise type, but also add a little more overhead to using (I expect people should just return this as-is pretty much)